Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs

```diff
@@ -312,11 +316,118 @@ public async Task Submit(SubmitSubmissionServiceModel model)
+        }
+
+        public void ProcessExecutionResult(SubmissionExecutionResult submissionExecutionResult)
+        {
+            var submission = this.submissionsData
+                .GetByIdQuery(submissionExecutionResult.SubmissionId)
+                .Include(s => s.Problem.Tests)
+                .FirstOrDefault();
+
+            if (submission == null)
+            {
+                throw new BusinessServiceException(
+                    $"Submission with Id: \"{submissionExecutionResult.SubmissionId}\" not found.");
+            }
+
+            var exception = submissionExecutionResult.Exception;
+            var executionResult = submissionExecutionResult.ExecutionResult;
+
+            submission.ProcessingComment = null;
+            this.testRunsDataService.DeleteBySubmission(submission.Id);
+
+            if (executionResult != null)
+            {
+                this.ProcessTestsExecutionResult(submission, executionResult);
+            }
+            else if (exception != null)
+            {
+                submission.ProcessingComment = exception.Message + Environment.NewLine + exception.StackTrace;
+            }
+            else
+            {
+                submission.ProcessingComment = "Invalid execution result received. Please contact an administrator.";
+            }
+        }
+
+        private void ProcessTestsExecutionResult(
+            Submission submission,
+            ExecutionResultResponseModel executionResult)
+        {
+            submission.IsCompiledSuccessfully = executionResult.IsCompiledSuccessfully;
+            submission.CompilerComment = executionResult.CompilerComment;
+            submission.Points = executionResult.TaskResult.Points;
+
+            if (!executionResult.IsCompiledSuccessfully)
+            {
+                this.UpdateResults(submission);
+                return;
+            }
+
+            var testResults = executionResult
+                                  .TaskResult
+                                  ?.TestResults
+                              ?? Enumerable.Empty<TestResultResponseModel>();
+
+            foreach (var testResult in testResults)
+            {
+                var resultType = (TestRunResultType)Enum.Parse(typeof(TestRunResultType), testResult.ResultType);
+
+                var testRun = new TestRun
+                {
+                    CheckerComment = testResult.CheckerDetails.Comment,
+                    ExpectedOutputFragment = testResult.CheckerDetails.ExpectedOutputFragment,
+                    UserOutputFragment = testResult.CheckerDetails.UserOutputFragment,
+                    ExecutionComment = testResult.ExecutionComment,
+                    MemoryUsed = testResult.MemoryUsed,
+                    ResultType = resultType,
+                    TestId = testResult.Id,
+                    TimeUsed = testResult.TimeUsed,
+                };
+
+                submission.TestRuns.Add(testRun);
+            }
```

Comment:
Always prefer `var result = collection.Select(...)` over `foreach(var obj in collection`;